### PR TITLE
[PA-1401] TC Automation - Delete failed and in-progress backup and restore by non-admin user from admin

### PR DESCRIPTION
**Pull request by kphalgun-px**
**What this PR does / why we need it**:
This PR adds a test case named "DeleteFailedInProgressBackupAndRestoreOfUserFromAdmin" to automate the use case of deleting failed and in-progress backup and restore by non-admin user from admin. This test case will increase the test coverage of Px-Backup TC Automation and ensure that the orphaned object feature functions as intended.

**Which issue(s) this PR fixes** (optional)
Closes #PA-1401

**Special notes for your reviewer**:
This test case is applicable to Px-Backup versions 2.6.0 and above.

TestRail: https://portworx.testrail.net/index.php?/cases/view/87564

Please review the following Jenkins builds:

v2.5.2

S3: I will update the results once the test run is complete.

NFS: I will update the results once the test run is complete.

### CODE REVIEW

**Comment by kphalgun-px** [comment]
The commit checks will fail because I neither set the upstream to px-backup/2.5.2-dev nor rebased with the topic/krishna-px/PA-1528-v7 branch, in order to keep the review process easy.

File: tests/backup/backup_share_test.go [hunk 1]
```diff
@@ -4215,3 +4215,400 @@ var _ = Describe("{SwapShareBackup}", func() {
 		wg.Wait()
 	})
 })
+
+// 1557-T1
```
Comment (kphalgun-px) [inline tests/backup/backup_share_test.go:4219]: This ID serves as my reference for tracking the test cases in the topic/krishna-px/PA-1528-v6 branch. I will remove this at the end of this review.

File: tests/backup/backup_share_test.go [hunk 2]
```diff
@@ -4215,3 +4215,400 @@ var _ = Describe("{SwapShareBackup}", func() {
...
+				for i := 0; i < numberOfBackups; i++ {
+					backupName := fmt.Sprintf("%s-%s-%d-%v", BackupNamePrefix, namespace, i, time.Now().Unix())
+					wg.Add(1)
+					go createBackup(backupName, namespace)
+					userInProgressBackupNames = append(userInProgressBackupNames, backupName)
+				}
+			}
+			wg.Wait()
+			log.Infof("The list of in progress user backups taken are: %v", userInProgressBackupNames)
+		})
```
Comment (kphalgun-px) [inline tests/backup/backup_share_test.go:4327]: I might be missing a step to verify backup by non-admin user is listed from admin pages while backup is in progress state.

File: tests/backup/backup_share_test.go [hunk 3]
```diff
@@ -4215,3 +4215,400 @@ var _ = Describe("{SwapShareBackup}", func() {
...
+				for i := 0; i < numberOfBackups; i++ {
+					backupName := fmt.Sprintf("%s-%s-%d-%v", BackupNamePrefix, namespace, i, time.Now().Unix())
+					wg.Add(1)
+					go createBackup(backupName, namespace)
+					userFailedBackupNames = append(userFailedBackupNames, backupName)
+				}
+			}
+			wg.Wait()
+			log.Infof("The list of failed user backups taken are: %v", userFailedBackupNames)
+		})
```
Comment (kphalgun-px) [inline tests/backup/backup_share_test.go:4425]: I might be missing a step to verify backup by non-admin user is listed from admin pages while backup is in failed state. @ak-px

File: tests/backup/backup_share_test.go [hunk 4]
```diff
@@ -4215,3 +4215,400 @@ var _ = Describe("{SwapShareBackup}", func() {
...
+				userInProgressRestoreNames = append(userInProgressRestoreNames, restoreName)
+			}
+			for backupName, namespace := range userSuccessfulBackupMap {
+				wg.Add(1)
+				restoreName := fmt.Sprintf("%s-%s", restoreNamePrefix, backupName)
+				go createRestore(backupName, restoreName, namespace)
+			}
+			wg.Wait()
+			log.Infof("The list of in progress user restores taken are: %v", userInProgressRestoreNames)
+		})
```
Comment (kphalgun-px) [inline tests/backup/backup_share_test.go:4489]: I might be missing a step to verify restore by non-admin user is listed from admin pages while restore is in inprogress state. @ak-px

File: tests/backup/backup_share_test.go [hunk 5]
```diff
@@ -4215,3 +4215,400 @@ var _ = Describe("{SwapShareBackup}", func() {
...
+				userFailedRestoreNames = append(userFailedRestoreNames, restoreName)
+			}
+			for backupName, namespace := range userSuccessfulBackupMap {
+				wg.Add(1)
+				restoreName := fmt.Sprintf("%s-%s", restoreNamePrefix, backupName)
+				go createRestore(backupName, restoreName, namespace)
+			}
+			wg.Wait()
+			log.Infof("The list of failed user restores taken are: %v", userFailedRestoreNames)
+		})
```
Comment (kphalgun-px) [inline tests/backup/backup_share_test.go:4586]: I might be missing a step to verify restore by non-admin user is listed from admin pages while restore is in failed state. @ak-px

File: tests/backup/backup_share_test.go [hunk 6]
```diff
@@ -4215,3 +4215,400 @@ var _ = Describe("{SwapShareBackup}", func() {
...
+				}
+			}
+			wg.Wait()
+			log.Infof("The list of in progress user backups taken are: %v", userInProgressBackupNames)
+		})
+		Step(fmt.Sprintf("Remove in progress backups taken by user %s from the admin", infraAdminUser), func() {
+			log.InfoD(fmt.Sprintf("Removing in progress backups taken by user %s from the admin", infraAdminUser))
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			backupEnumerateRequest := &api.BackupEnumerateRequest{
```
Comment (ak-px) [inline tests/backup/backup_share_test.go:4332]: You can add the ownerid to the enumerate options.

File: tests/backup/backup_share_test.go [hunk 7]
```diff
@@ -4215,3 +4215,400 @@ var _ = Describe("{SwapShareBackup}", func() {
...
+					dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying creation of backup %s of namespace %s", backupName, namespace))
+					break
+				}
+			}
+			for _, namespace := range appNamespaces {
+				for i := 0; i < numberOfBackups; i++ {
+					backupName := fmt.Sprintf("%s-%s-%d-%v", BackupNamePrefix, namespace, i, time.Now().Unix())
+					wg.Add(1)
+					go createBackup(backupName, namespace)
+					userInProgressBackupNames = append(userInProgressBackupNames, backupName)
```
Comment (ak-px) [inline tests/backup/backup_share_test.go:4322]: we no need to store here as list for user backups, instead fetch the backups using ownerid before you go for deletion.

File: tests/backup/backup_share_test.go (hunk 2 above)
Comment (ak-px) [reply tests/backup/backup_share_test.go:4327]: @kphalgun-px , As we discussed before it's an important step for orphaned backup delete. Ive added a comment where you need to add it.

File: tests/backup/backup_share_test.go [hunk 8]
```diff
@@ -4215,3 +4215,400 @@ var _ = Describe("{SwapShareBackup}", func() {
...
+				}
+			}
+			wg.Wait()
+			log.Infof("The list of failed user backups taken are: %v", userFailedBackupNames)
+		})
+		Step(fmt.Sprintf("Delete failed backups taken by user %s from the admin", infraAdminUser), func() {
+			log.InfoD(fmt.Sprintf("Deleting failed backups taken by user %s from the admin", infraAdminUser))
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			for _, backupName := range userFailedBackupNames {
```
Comment (ak-px) [inline tests/backup/backup_share_test.go:4430]: Fetch using ownerID here as well.

File: tests/backup/backup_share_test.go (hunk 4 above)
Comment (ak-px) [reply tests/backup/backup_share_test.go:4489]: @kphalgun-px , owner id for restore is not supported in enumerate options, you can drop listing for restore deletion steps

File: tests/backup/backup_share_test.go (hunk 5 above)
Comment (ak-px) [reply tests/backup/backup_share_test.go:4586]: @kphalgun-px , owner id for restore is not supported in enumerate options, you can drop listing for restore deletion steps

File: tests/backup/backup_share_test.go (hunk 8 above)
Comment (kphalgun-px) [reply tests/backup/backup_share_test.go:4430]: Using admin ctx to fetch the backup UID as suggested by @sn-px.

File: tests/backup/backup_share_test.go (hunk 2 above)
Comment (kphalgun-px) [reply tests/backup/backup_share_test.go]: Done.

File: tests/backup/backup_share_test.go (hunk 3 above)
Comment (kphalgun-px) [reply tests/backup/backup_share_test.go]: Done.

File: tests/backup/backup_share_test.go (hunk 4 above)
Comment (kphalgun-px) [reply tests/backup/backup_share_test.go]: Done. Used GetAllRestoreNamesByOwnerID in topic/krishna-px/PA-1528-v7.

File: tests/backup/backup_share_test.go (hunk 5 above)
Comment (kphalgun-px) [reply tests/backup/backup_share_test.go]: Done. Used GetAllRestoreNamesByOwnerID in topic/krishna-px/PA-1528-v7.

File: tests/backup/backup_share_test.go (hunk 6 above)
Comment (kphalgun-px) [reply tests/backup/backup_share_test.go]: Done.

File: tests/backup/backup_share_test.go (hunk 7 above)
Comment (kphalgun-px) [reply tests/backup/backup_share_test.go]: I used to check if the in progress backups are listed using admin ctx.

File: tests/backup/backup_orphaned_delete_test.go [hunk 9]
```diff
@@ -0,0 +1,505 @@
...
+			backupNamesByOwnerID, err := GetAllBackupNamesByOwnerID(userOwnerID, orgID, ctx)
+			log.FailOnError(err, "failed to fetch backup names with owner id %s from the admin", userOwnerID)
+			for _, backupName := range userFailedBackupNames {
+				if !IsPresent(backupNamesByOwnerID, backupName) {
+					err := fmt.Errorf("failed backup %s is not listed in backup names %s", backupName, backupNamesByOwnerID)
+					log.FailOnError(fmt.Errorf(""), err.Error())
+				}
+			}
+		})
+		Step(fmt.Sprintf("Delete failed backups taken by user %s from the admin", infraAdminUser), func() {
```
Comment (snigdha-px) [inline tests/backup/backup_orphaned_delete_test.go:266]: Can we combine this and the previous step i.e check if the backup is present -> if yes -> delete

File: tests/backup/backup_orphaned_delete_test.go [hunk 10]
```diff
@@ -0,0 +1,505 @@
...
+				defer mu.Unlock()
+				userInProgressRestoreNames = append(userInProgressRestoreNames, restoreName)
+			}
+			for backupName, namespace := range userSuccessfulBackupMap {
+				wg.Add(1)
+				restoreName := fmt.Sprintf("%s-%s", restoreNamePrefix, backupName)
+				go createRestore(backupName, restoreName, namespace)
+			}
+			wg.Wait()
+			log.Infof("The list of in progress user restores taken are: %v", userInProgressRestoreNames)
```
Comment (snigdha-px) [inline tests/backup/backup_orphaned_delete_test.go:345]: How are we making sure that the restore is in progress ?

File: tests/backup/backup_orphaned_delete_test.go [hunk 11]
```diff
@@ -0,0 +1,505 @@
...
+		Step(fmt.Sprintf("Delete failed restores taken by user %s from the admin", infraAdminUser), func() {
+			log.InfoD(fmt.Sprintf("Deleting failed restores taken by user %s from the admin", infraAdminUser))
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			for _, restoreName := range userFailedRestoreNames {
+				restoreUID, err := Inst().Backup.GetRestoreUID(ctx, restoreName, orgID)
+				log.FailOnError(err, "failed to fetch restore %s uid", restoreName)
+				err = DeleteRestoreWithUID(restoreName, restoreUID, orgID, ctx)
+				log.FailOnError(err, "failed to delete restore %s uid", restoreName)
+			}
```
Comment (snigdha-px) [inline tests/backup/backup_orphaned_delete_test.go:487]: same.
Can we combine the two steps

File: tests/backup/backup_orphaned_delete_test.go (hunk 9 above)
Comment (kphalgun-px) [reply tests/backup/backup_orphaned_delete_test.go:266]: In TestRail these are two different steps.

File: tests/backup/backup_orphaned_delete_test.go (hunk 10 above)
Comment (kphalgun-px) [reply tests/backup/backup_orphaned_delete_test.go:345]: The CreateRestoreWithoutCheck function will not wait for the restores to be successful. I am ensuring that the restores are in in-progress while deleting them.

File: tests/backup/backup_orphaned_delete_test.go (hunk 11 above)
Comment (kphalgun-px) [reply tests/backup/backup_orphaned_delete_test.go:487]: In TestRail these are two different steps.

**Review by ak-px — APPROVED** [review]
LGTM

**Review by snigdha-px — APPROVED** [review]
LGTM

**Comment by kphalgun-px** [comment]
Thank you for the reviews. These changes are now included in PR #1663.

**Pull request closed without merging**